src/hotspot/share/opto/addnode.cpp

```diff
@@ -395,9 +395,164 @@ Node* AddNode::IdealIL(PhaseGVN* phase, bool can_reshape, BasicType bt) {
+// generalized pattern `(a*b) + (a*c)` into `a*(b + c)` is handled by AddNode::IdealIL().
+Node* AddNode::convert_serial_additions(PhaseGVN* phase, BasicType bt) {
+  // We need to make sure that the current AddNode is not part of a MulNode that has already been optimized to a
+  // power-of-2 addition (e.g., 3 * a => (a << 2) + a). Without this check, GVN would keep trying to optimize the same
+  // node and can't progress. For example, 3 * a => (a << 2) + a => 3 * a => (a << 2) + a => ...
+  if (find_power_of_two_addition_pattern(this, bt).valid) {
+    return nullptr;
+  }
+
+  Node* in1 = in(1);
+  Node* in2 = in(2);
+
+  // (1) Simple addition pattern (e.g., a + a)
+  Multiplication mul = find_simple_addition_pattern(in1, bt);
+
+  // (2) Simple lshift pattern (e.g., a << CON)
+  if (!is_valid_multiplication(mul, in2)) {
+    mul = find_simple_lshift_pattern(in1, bt);
+  }
+
+  // (3) Simple multiplication pattern (e.g., CON * a)
+  if (!is_valid_multiplication(mul, in2)) {
+    mul = find_simple_multiplication_pattern(in1, bt);
+  }
+
+  // (4) Power-of-two addition pattern (e.g., (a << CON1) + (a << CON2))
+  // While multiplications can be potentially optimized to power-of-2 subtractions (e.g., a * 7 => (a << 3) - a),
+  // (x - y) + y => x is already handled by the Identity() methods. So, we don't need to check for that pattern here.
+  if (!is_valid_multiplication(mul, in2)) {
+    mul = find_power_of_two_addition_pattern(in1, bt);
+  }
+
+  // We've tried everything.
+  if (!is_valid_multiplication(mul, in2)) {
+    return nullptr;
+  }
+
+  Node* con = (bt == T_INT)
+              ? (Node*) phase->intcon(java_add((jint) mul.multiplier, (jint) 1)) // Overflow at max_jint
+              : (Node*) phase->longcon(java_add((jlong) mul.multiplier, (jlong) 1));
+  return MulNode::make(con, mul.variable, bt);
+}
+
+// Try to match `a + a`. On success, return a struct with `.valid = true`, `variable = a`, and `multiplier = 2`.
+// The method matches `n` for pattern: AddNode(a, a).
+AddNode::Multiplication AddNode::find_simple_addition_pattern(const Node* n, BasicType bt) {
+  if (n->Opcode() == Op_Add(bt) && n->in(1) == n->in(2)) {
+    return Multiplication{true, n->in(1), 2};
+  }
+
+  return Multiplication{};
+}
+
+// Try to match `a << CON`. On success, return a struct with `.valid = true`, `variable = a`, and
+// `multiplier = 1 << CON`.
+// Match `n` for pattern: LShiftNode(a, CON).
+// Note that the power-of-2 multiplication optimization could potentially convert a MulNode to this pattern.
+AddNode::Multiplication AddNode::find_simple_lshift_pattern(const Node* n, BasicType bt) {
+  // Note that power-of-2 multiplication optimization could potentially convert a MulNode to this pattern
+  if (n->Opcode() == Op_LShift(bt) && n->in(2)->is_Con()) {
+    Node* con = n->in(2);
+    if (!con->is_top()) {
+      return Multiplication{true, n->in(1), java_shift_left(1, con->get_int(), bt)};
+    }
+  }
+
+  return Multiplication{};
```

Comment:
```suggestion
  return Multiplication::make_invalid();
```
I would prefer wrapping this in something that is immediately clear at the call site, so the reader does not have to go look at the internals of `Multiplication`.